src/odemis/driver/piezomotor.py

```diff
@@ -0,0 +1,453 @@
+Foundation.
+
+Odemis is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
+PURPOSE. See the GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License along with
+Odemis. If not, see http://www.gnu.org/licenses/.
+'''
+from __future__ import division
+
+import glob
+import logging
+from odemis import model
+from odemis.model import Actuator, CancellableThreadPoolExecutor
+from odemis.util import to_str_escape
+import os
+import threading
+import time
+import serial
+from threading import Thread
+import copy
+from odemis.model import Axis, HwComponent
+
+BAUDRATE = 115200
+WFM_STEPSIZE = 5e-6  # m / step
+USTEPS_PER_WFM = 8192
+ENCODER_STEPSIZE = 1e-9  # m / step
+
+EOL = b'\r'  # 0x0D, carriage return
+
+# Waveforms
+WAVEFORM_RHOMB = 1  # fast max speed
+WAVEFORM_DELTA = 2  # preferred, higher accuracy
+WAVEFORM_PARK = 4  # power off
+
+
+class PMAxis(Axis, HwComponent):
+    def __init__(self, name, role, canAbs=True, choices=None, unit=None,
+                 range=None, speed=None, canUpdate=False, *args, **kwargs):
+        HwComponent.__init__(self, name, role, *args, **kwargs)
+        Axis.__init__(self, canAbs=canAbs, choices=choices, unit=unit,
+                      range=range, speed=speed, canUpdate=canUpdate)
+
+
+class PMD401(Actuator):
+    """
+    This represents the PMD401 motor controller for the PiezoMotor LEGS motors.
+    The specification for the hardware interface can be found in the document
+     "PiezoMotor_PMD401_Technical_Manual.pdf".
+
+    The class provides only part of the functionality of the model.Actuator interface. It fully
+    implements the functions moveRel and stop. However, there is no reference function and
+    moveAbs is only possible if we are operating in closed loop, i.e. an encoder (stage position sensor)
+    is present.
+    """
+
+    def __init__(self, name, role, port, axes, ax_numbers, inverted=None, closed_loop=True, speed=1e-6, **kwargs):
+        """
+        :arg ax_numbers (dict: str --> int): axis name --> axis number
+        :arg closed loop (bool): True if stage position sensor is present, False otherwise
+        :arg speed (float): speed in m/s
+        """
+        # It is also possible to send a command to the device that lets the firmware deal with
+        # inverted axes directly, but since we already have code for this in odemis, we'll
+        # invert the axes in the software.
+        Actuator.__init__(self, name, role, axes=axes, inverted=inverted, **kwargs)
+        self._executor = CancellableThreadPoolExecutor(max_workers=1)  # one task at a time
+        self._ser_access = threading.RLock()
+
+        self._ax_numbers = ax_numbers
+        self.closed_loop = closed_loop
+        self._speed = speed
+
+        # Connect to hardware
+        self._serial = self._findDevice(port)
+
```

Comment:
You need to provide the `.position` VA, updated at least after each move.
You should also provide a `.referenced` VA, which at least should be empty.